src/devices/SoftPwm/SoftwarePwmChannel.cs

```diff
@@ -105,106 +110,89 @@ public SoftwarePwmChannel(int pinNumber, int frequency = 400, double dutyCycle =
                 _shouldDispose = shouldDispose;
             }
 
-            _pin = pinNumber;
-            _controller.OpenPin(_pin, PinMode.Output);
-            _usePrecisionTimer = usePrecisionTimer;
-            _isRunning = false;
+            UpdatePulseWidthParameters();
 
-            _frequency = frequency;
-            _dutyCycle = dutyCycle;
+            _thread = new Thread(Run);
 
-            UpdatePulseWidthParameters();
+            if (usePrecisionTimer)
+            {
+                _stopwatch = new Stopwatch();
+                _stopwatch.Start();
+                _thread.Priority = ThreadPriority.Highest;
+            }
+
+            _controller.OpenPin(_pin, PinMode.Output);
 
-            _runningThread = new Thread(RunSoftPWM);
-            _runningThread.Start();
+            _thread.Start();
         }
 
         private void UpdatePulseWidthParameters()
         {
-            _periodInMilliseconds = 1000.0 / _frequency;
-            _pulseWidthInMilliseconds = _dutyCycle * _periodInMilliseconds;
+            double cycleTicks = TimeSpan.TicksPerSecond / (double)_frequency;
+
+            double pinHighTicks = cycleTicks * _dutyCycle;
+            _pinHighTime = TimeSpan.FromTicks((long)pinHighTicks);
+
+            double pinLowTicks = cycleTicks - pinHighTicks;
+            _pinLowTime = TimeSpan.FromTicks((long)pinLowTicks);
         }
 
-        private void RunSoftPWM()
+        private void Run()
         {
-            if (_usePrecisionTimer)
+            while (!_isTerminating)
             {
-                Thread.CurrentThread.Priority = ThreadPriority.Highest;
-            }
-
-            while (_runThread)
-            {
-                // Write the pin high for the appropriate length of time
-                if (_isRunning)
+                if (!_isRunning)
                 {
-                    if (_pulseWidthInMilliseconds != 0)
-                    {
-                        _controller.Write(_pin, PinValue.High);
-                        _isStopped = false;
-                    }
-
-                    // Use the wait helper method to wait for the length of the pulse
-                    if (_usePrecisionTimer)
-                    {
-                        Wait(_pulseWidthInMilliseconds);
-                    }
-                    else
-                    {
-                        Task.Delay(TimeSpan.FromMilliseconds(_pulseWidthInMilliseconds)).Wait();
-                    }
-
-                    // The pulse if over and so set the pin to low and then wait until it's time for the next pulse
-                    _controller.Write(_pin, PinValue.Low);
+                    Thread.Yield();
```

Comment:
The old logic just wasted cycles in the background thread when the PWM wasn't running. This makes it a good citizen and yields until the user asks to start running again.

Reply:
The old logic was also constantly turning the pin off, which isn't necessarily desirable if the pin is also being controlled by other logic.